components/eam/src/physics/cam/zm_conv.F90

```diff
@@ -2688,27 +3797,133 @@ subroutine cldprp(lchnk   , &
       end do
    end do
 
+   do i = 1,il2g
+     if (zm_microp) then
+        tmplel(i) = jlcl(i)+1
+     else
+        tmplel(i) = jb(i)
+     end if
+   end do
+
 ! compute condensation in updraft
    do k = pver,msg + 2,-1
       do i = 1,il2g
-         if (k >= jt(i) .and. k < jb(i) .and. eps0(i) > 0._r8) then
+        if (k >= jt(i) .and. k < tmplel(i) .and. eps0(i) > 0._r8) then
+          if (zm_microp) then
+             cu(i,k) = ((mu(i,k)*su(i,k)-mu(i,k+1)*su(i,k+1))/ &
+                     dz(i,k)- eu(i,k)*s(i,k)+du(i,k)*su(i,k))/(rl/cp)  &
+                       - latice*frz(i,k)/rl
+          else
+
             cu(i,k) = ((mu(i,k)*su(i,k)-mu(i,k+1)*su(i,k+1))/ &
                       dz(i,k)- (eu(i,k)-du(i,k))*s(i,k))/(rl/cp)
+          end if
             if (k == jt(i)) cu(i,k) = 0._r8
             cu(i,k) = max(0._r8,cu(i,k))
          end if
       end do
    end do
 
+   if (zm_microp) then
+
+      tug(:il2g,:) = t(:il2g,:)
+      fice(:,:)    = 0._r8
+
+      do k = pver, msg+2, -1
+         do i = 1, il2g
+            tug(i,k) = su(i,k) - grav/cp*zf(i,k)
+         end do
+      end do
+
+      do k = 1, pver-1
+         do i = 1, il2g
+
+            if (tug(i,k+1) > 273.15_r8) then
+               ! If warmer than tmax then water phase
+               fice(i,k) = 0._r8
+
+            else if (tug(i,k+1) < 233.15_r8) then
+               ! If colder than tmin then ice phase
+               fice(i,k) = 1._r8
+
+            else
+               ! Otherwise mixed phase, with ice fraction decreasing linearly
+               ! from tmin to tmax
+               fice(i,k) =(273.15_r8 - tug(i,k+1)) / 40._r8
+            end if
+         end do
+      end do
+
+      do k = 1, pver
+         do i = 1,il2g
+            loc_microp_st%cmei(i,k) = cu(i,k)* fice(i,k)
+            loc_microp_st%cmel(i,k) = cu(i,k) * (1._r8-fice(i,k))
+         end do
+      end do
+
+      call  zm_mphy(su,   qu,    mu,    du,   eu,  loc_microp_st%cmel,  loc_microp_st%cmei,   zf,    p,   & 
```

Comment:
Couldn't we just pass the loc_microp_st variable as a single argument to clean up this interface?

Reply:
To do so, we need to rewrite zm_microphysics.F90 and would not meet the deadline.